rfd/0033-sni-alpn-teleport-proxy-routing.md

```diff
@@ -0,0 +1,160 @@
+
+```
+                                                                                                                                            +------------+                                 
+                                                                                                                                            |            |                                 
+                                                                                                                                            |    Auth    |                                 
+                                                                                                                                            |            |                                 
+                                                                                                                                            +--^------|--+                                 
+                                                                                                                                               |      |                                    
+                                                                                                                                               |      |                                    
+                                                                                                                                        5) Auth Req 6) Auth Resp                           
+                                                                                                                                               |      |                                    
+                                               teleport-db-protocol                                                                            |      |                                    
++-----------+  1) DB Protocol     +-----------+               +-----------------+  3) DB Protocol  +-------------------+   4) DB Protocol   +--|------v--+  7) DB Protocol+---------------+
+|           ---------------------->           ---------------->                 ------------------->                   --------------------->            ----------------->               |
+| DB Client |  12) DB Protocol    | Local DB  | 11) TLS       | Remote Teleport |  10) DB Protocol |  DB Proxy Service |   9) DB Protocol   | DB Service |  8) DB Protocol|  DB Instance  |
+|           <----------------------   Proxy   <----------------     Proxy       <-------------------                   <---------------------            <-----------------               |
++-----------+                     +-----------+               +-----------------+                  +-------------------+                    +------------+                +---------------+
+```
+
+#### tsh proxy start db db-instance-name [-p port] 
```

Comment:
Do we really need a dedicated command for db proxy? Can we just have a command to start a local proxy that's protocol-agnostic? `tsh proxy start` for example - it would be just receiving TLS connections from local clients and dispatching according to SNI/ALPN info.

Reply:
That will require detecting what protocol client use. In case of MySQL protocol where server sends first message to the client to initialise connection  detecting MySQL protocol seems to be not possible.